Blocks: Reorganize files and directories

coreymckrill opened the pull request:
This reorganizes the JS files for blocks into a structure that is a bit more logical and more closely matches the organization found in the Gutenberg plugin. It attempts to address the issues discussed in #47 and pretty much follows the pattern suggested [in this comment](https://github.com/WordPress/wordcamp.org/issues/47#issuecomment-485598737)

```
blocks
    build
        blocks.min.css
        blocks.min.js
    source
        block
            organizers
            sessions
            speakers
            sponsors
        component
            block-content
            block-controls
            image
            item-select
            post-list
        data
        i18n
        blocks.js
```

coreymckrill committed:
Rearrange files and directories, update references

coreymckrill committed:
Refactor away from unnecessary class inheritance

The `BlockControls` class was originally built solely in order to
provide a method for accessing a mode option's label in a DRY way.
However, it's a poor use of class inheritance, and the method is
better anyway as an exportable function that can work to get the label
of any option.

coreymckrill committed:
Delete build folder and built files

I accidentally committed these when reorganizing the file structure.
These should be ignored by Git but committed (at least for now) to
SVN.

coreymckrill committed:
Organizers: Eliminate default exports

See https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad

iandunn commented on public_html/wp-content/mu-plugins/blocks/source/block/index.js:
```diff
@@ -0,0 +1,11 @@
+import * as organizers from './organizers';
```
`SHOULD` - Should the parent folder be named `blocks/` (plural), since it contains multiple, and for consistency with G? Same for `components/`, `views/`, the `BLOCK` constant, etc.

coreymckrill replied:
JS folders renamed here: 116c34c9a2b120c045d89964db1e6af8c6bc9465

coreymckrill replied:
PHP folders renamed here: dc2a6650bb6b7d8e8b87abdaba951f644ec64ee6

iandunn commented on public_html/wp-content/mu-plugins/blocks/source/block/index.js:
```diff
@@ -0,0 +1,11 @@
+import * as organizers from './organizers';
```
`SHOULD` - Normally I'd definitely prefer `source` to `src`, but in this case I'm wondering if the grandparent folder be  `src`, to match what `wp-scripts` expects? That'll make it easier to eventually migrate to that instead of our custom configs (assuming that's possible).

vedanshujain replied:
I'd actually prefer `src`, because what I see more commonplace. 
I'm curious, can `wp-scripts` could be configured to use with `source`? If so then might as well keep it this way.

coreymckrill replied:
Aw man, I picked `source` because I remembered that you preferred spelling things out! 😆 

It's unclear to me right now if/how we would migrate to a pure usage of `wp-scripts` instead of our own config, but if we did, I think it would be easy to change it to `src` at that time if necessary. There are very few places in the code that need to reference that part of the path.

iandunn replied:
>  can wp-scripts could be configured to use with source?

I think so, but it'd require a custom webpack config, which would default a lot of the purpose of using it. There may be a way to do it with simple npm script parameters or something, though.

> It's unclear to me right now if/how we would migrate to a pure usage of wp-scripts instead of our own config, but if we did, I think it would be easy to change it to src at that time if necessary. 

Good point, I withdrawal my objection :)

iandunn commented on public_html/wp-content/mu-plugins/blocks/blocks.php:
```diff
@@ -52,14 +52,14 @@ function register_block_categories( $default_categories ) {
 function register_assets() {
```
`SHOULD` - Should we rename `includes/shared/` to `includes/components` too, so it's consistent w/ the JS? Maybe move the block PHP files into `includes/blocks` for the same reason?

coreymckrill replied:
PHP stuff reorganized here: dc2a6650bb6b7d8e8b87abdaba951f644ec64ee6

iandunn commented on public_html/wp-content/mu-plugins/blocks/source/block/sessions/block-content.js:
```diff
@@ -13,11 +13,11 @@ const { __ }        = wp.i18n;
+import { FeaturedImage }                                             from '../../component/image';
+import { PostList }                                                  from '../../component/post-list';
+import { filterEntities }                                            from '../../data';
+import { tokenSplit, arrayTokenReplace, intersperse, listify }       from '../../i18n';
```
`SHOULD` - I wonder if `helpers/i18n.js` would be more future-proof? We probably won't have more custom i18n stuff, but it seems somewhat likely that we'll have more generic helpers.

I'm open to other terms besides "helpers" too, that was just the first that came to mind.

coreymckrill replied:
This follows the convention of Gutenberg, which also has a `i18n` package. I don't think it's a problem to have this separate directory with only a few functions in it. In fact, I think that having more, explicitly named folders will make it easier for people to browse the code and see what kinds of "helper" functions are available.

iandunn replied:
👍

iandunn commented on public_html/wp-content/mu-plugins/blocks/blocks.php:
```diff
@@ -52,14 +52,14 @@ function register_block_categories( $default_categories ) {
 	wp_register_script(
 		'wordcamp-blocks',
-		PLUGIN_URL . 'assets/blocks.min.js',
+		PLUGIN_URL . 'build/blocks.min.js',
```
I like `blocks.min.js` better than `index.min.js` 👍

iandunn commented on public_html/wp-content/mu-plugins/blocks/source/block/organizers/index.js:
```diff
@@ -6,22 +6,22 @@ const { __ } = wp.i18n;
-	title       : __( 'Organizers',                'wordcamporg' ),
+export const SETTINGS = {
+	title       : __( 'Organizers', 'wordcamporg' ),
 	description : __( 'Add a list of organizers.', 'wordcamporg' ),
```
`SHOULD` - Why remove the alignment here?

coreymckrill replied:
This made it consistent with the other blocks 🤷‍♂

iandunn replied:
👍 We can fix them all later

iandunn commented on public_html/wp-content/mu-plugins/blocks/source/block/sessions/block-content.js:
```diff
@@ -13,11 +13,11 @@ const { __ }        = wp.i18n;
+import { ItemTitle, ItemHTMLContent, ItemPermalink, BlockNoContent } from '../../component/block-content';
+import { FeaturedImage }                                             from '../../component/image';
+import { PostList }                                                  from '../../component/post-list';
+import { filterEntities }                                            from '../../data';
```
I like having a separate folder for `data` 👍 

`MIGHT` - I wonder if `stores` or `data-stores` would be a more descriptive name, since `data` is very generic. I can see it fitting in this context, though.

coreymckrill replied:
`data` is what the data stores package is called in Gutenberg, so 🤷‍♂

iandunn replied:
🤷‍♂

iandunn commented on public_html/wp-content/mu-plugins/blocks/source/block/sessions/index.js:
```diff
@@ -6,22 +6,22 @@ const { __ } = wp.i18n;
 	category    : 'wordcamp',
 	supports    : supports,
-	edit        : edit,
+	edit        : Edit,
```
Nice work making things like this consistent ❤️

iandunn commented on public_html/wp-content/mu-plugins/blocks/source/block/speakers/block-content.js:
```diff
@@ -15,11 +15,11 @@ const { escapeAttribute }     = wp.escapeHtml;
+import { AvatarImage }                                               from '../../component/image';
+import { PostList }                                                  from '../../component/post-list';
+import { filterEntities }                                            from '../../data';
+import { tokenSplit, arrayTokenReplace }                             from '../../i18n';
```
❤️ alphabetizing :)

iandunn commented on public_html/wp-content/mu-plugins/blocks/source/blocks.js:
```diff
@@ -0,0 +1,13 @@
+ */
+import { BLOCKS } from './block';
+
+BLOCKS.forEach( ( { NAME, SETTINGS } ) => {
```
`SHOULD` - When these are passed to the closure, it seems like they're no longer constants, but instead local vars that would be camelCase.

coreymckrill replied:
Not sure what you're suggesting here. They are defined as constants, which is why I made them all caps (though we are not completely consistent in the codebase with this). When they are imported and deconstructed, the case of the variable names still applies, similar to how we're importing `LABEL` and `ICON` in various places. Are you suggesting we rename them during deconstruction, i.e. `{ NAME: name, SETTINGS: settings }`? That seems unnecessary to me...

iandunn replied:
er, yeah, good point, that would be kind of annoying. it's fine to leave it.

iandunn replied:
Although, I don't think we should uppercase local vars that are declared with `const` instead of `let`, only "global" ones, like these.

coreymckrill replied:
Ah, that's a good distinction.

iandunn commented on public_html/wp-content/mu-plugins/blocks/source/block/organizers/block-content.js:
```diff
@@ -19,7 +19,7 @@ import { filterEntities }                             from '../../data';
  * Component for displaying the block content.
  */
-class OrganizersBlockContent extends Component {
+export class BlockContent extends Component {
```
❤️ 

I'm guessing you're planning to do this for the rest of the blocks/components too?

coreymckrill replied:
Yep, just committed the other blocks as well.

iandunn commented on public_html/wp-content/mu-plugins/blocks/source/block/organizers/inspector-controls.js:
```diff
@@ -2,7 +2,7 @@
  */
 const { PanelBody, PanelRow, SelectControl, ToggleControl } = wp.components;
-const { InspectorControls }                                 = wp.editor;
+const { InspectorControls: Inspector }                      = wp.editor;
```
🤔 

`SHOULD` - This feels kinda wrong to me. If there's a conflict, it seems like our custom components should be the ones be named more descriptively, rather than essential G components.

I'm not entirely sure either way, though.

coreymckrill replied:
I thought about this too. I could flip it so ours is the one named just `Inspector`, but then I'd probably want to rename the file to just `inspector.js` as well.

On the other hand, the Gutenberg element really is the whole Inspector Sidebar, while ours is just a component that adds some additional inspector controls, so semantically this makes more sense to me, even if we're renaming the Core comonent (which, yes, is due to a naming conflict)

iandunn replied:
🤔 

Well, to further complicate things, to me `Inspector` makes it sound like it's the entire `BlockInspector`, rather than just the controls for it, so I'm not sure that that name works well for either `wp.editor.InspectorControls` or our `OrganizerInspectorControls`.

iandunn replied:
If it'd be more practical/semanic to rename `wp.editor.InspectorControls` rather than our custom components, then maybe it should be named `BlockInspectorControls` ?

That still feels odd to me, though, because I perceive the platform's components as (conceptually) immutable, whereas the custom stuff we build on top of that is malleable, and created in light of that platform's prior existence. 

To draw an analogy to PHP, I don't think we'd ever to this:

```php
<?php

namespace WordCamp\Foo;
use function wp_insert_post as insert_post;

function wp_insert_post( $arr ) {
    // do some custom wrapper processing on $arr
    insert_post( $arr );
}
```

...instead, we'd just

```php
<?php

namespace WordCamp\Foo;

function insert_post( $arr ) {
    // do some custom wrapper processing on $arr
    wp_insert_post( $arr );
}
```

iandunn replied:
<a href="https://www.commitstrip.com/en/2015/10/27/one-of-the-coders-hardest-problems/"><img src="https://www.commitstrip.com/wp-content/uploads/2015/10/Strip-Trouver-le-nom-de-variable-english650-final.jpg" alt="There are only 2 hard things in computer science..." /></a>

😉

coreymckrill replied:
💯

coreymckrill replied:
Good point that platform component names should probably be immutable. What about eschewing "Inspector" altogether and going with something like `BlockSidebar`? Same idea, and it still follows the convention we're using with `BlockControls` and `BlockContent`...

iandunn replied:
🤔 I don't love it, since "inspector" feels more specific and meaningful.

🤔, maybe `InspectorControlContents` ? It's a bit of a mouthful, but kinda fits. It's not specific to Organizers, but I guess the intent is to make it more generic inside the component anyway. We could even shorter it to `ControlContents` since the "inspector" part is mostly obvious from it being wrapped in an `<InspectorControl>`. 

🤔 I don't love that either, and it could also conflict if we have other types of controls inside the parent component.

🤷‍♂Oh well, it's a crapshoot, pick the one you like best and :shipit:  😄

coreymckrill replied:
Maybe not a popular decision, but a decision nonetheless: 8ec11263bd5c149d52dc241d398cc6cecde5e2c1

coreymckrill committed:
Sessions: Eliminate default exports

See https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad

iandunn commented on public_html/wp-content/mu-plugins/blocks/source/block/organizers/block-controls.js:
```diff
@@ -7,28 +7,31 @@ import classnames from 'classnames';
  * WordPress dependencies
```
❤️

coreymckrill replied:
I can't tell what you're heart-ing here. I'll just take it as a general sign of encouragement 😁

iandunn replied:
Er, yeah, it makes more sense if you [view it in the context of the commit](https://github.com/WordPress/wordcamp.org/pull/117/commits/e9bcd2897aff6b1682a68496592134049d1d0e40), which is what I was looking at when I wrote it, but even then it's not entirely clear.

I was referring to the commit message (or, really, to the intent behind the commit, not the literal message, but that was fine too). It doesn't really apply to a specific line, but GitHub doesn't always let me comment on the commit as a whole (haven't noticed the pattern to when it does and doesn't, maybe it's only enabled for ones that are merged?)

Anyway, I should have quoted the commit message before my reply to make that clear :)

iandunn reviewed (changes requested):
Looks great overall. Left some comments that we'll want to resolve before merging.

coreymckrill committed:
Speakers: Eliminate default exports

See https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad

coreymckrill committed:
Sponsors: Eliminate default exports

See https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad

vedanshujain commented on public_html/wp-content/mu-plugins/blocks/source/block/organizers/block-controls.js:
```diff
@@ -7,49 +7,52 @@ import classnames from 'classnames';
+import { getOptionLabel }          from '../../component/item-select';
+import { BlockContent }            from './block-content';
+import { Select }                  from './select';
+import { LABEL }                   from './index';
```
Generic names here looks way cleaner, but it could cause problem down the line if need to grep for stuff, or if we have conflict with another library. Although I don't feel strongly about this so feel free to keep however :) .

vedanshujain replied:
That said, perhaps we can think about renaming `Select` to something else. Its way to generic IMO.

coreymckrill replied:
Fair point. What about `ContentSelect`?

coreymckrill replied:
Renamed to `ContentSelect`: c89a50f75be17fca4e69e78f98867e03ccd523c5

vedanshujain reviewed (approved):
Left some pretty minor comments but this looks awesome 👍

coreymckrill committed:
Rename some JS folders and update references

* block -> blocks
* component -> components

This makes them more consistent with naming in Gutenberg.

coreymckrill committed:
Reorganize PHP files, update namespaces

* Moved block includes into `blocks` subfolder.
* Moved component includes into `components` subfolder.
* Some functions in content.php were really components, so moved them
  to a new file in the `components` subfolder and updated their
  namespace.
* There are currently only three functions that seem like "utilities"
  or "helpers" that aren't components, so moved them to a file in a
  `utilities` subfolder and updated their namespace.
* Definitions seem fairly foundational to our whole blocks module,
  rather than some kind of specialized utility, so left them in the
  root includes folder.

coreymckrill committed:
Add missing dependency in PHP

coreymckrill committed:
Rename blocks' `Select` component to `ContentSelect`

`Select` was too generic.

coreymckrill committed:
Rename `InspectorControls` import to `CoreInspectorControlsContainer`

Naming is really hard.

The problem here is that "inspector controls" is really most accurate
name for the component within our block, but it's also the name used
by Core/Gutenberg for the _container_ that holds all the inspector
controls.

@iandunn suggested that components that we import from the platform
should be immutable, i.e. shouldn't be renamed, and initially I
agreed. However, ultimately I decided a rule like that was imposing
unnecessary limits on us by allowing code that we didn't write to make
it harder to accurately describe the contents of the code that we _did_
write. In other words, I think it was a poor naming choice for the Core
component.

Changing the name of an import can actually be helpful if the new
name makes it _more_ clear what the imported component is and does,
especially in the context of the code that is importing and using it.
Originally I had renamed the import simply `Inspector`, but that
definitely did not add clarity. `CoreInspectorControlsContainer` on the
other hand, while fairly verbose, makes it clear that A) it's imported
from Core, and B) it's actually a container for controls, not the
controls themselves.

coreymckrill commented:
Committed: https://meta.trac.wordpress.org/changeset/8779
